Blog about Azure Provider 2.0 (#2629)
pulumi/docs: content/blog/pulumi-azure-2-0/_index.md

## New Resources for Virtual Machine & Virtual Machine Scale Set 

There are four new resources for Virtual Machine & Virtual Machine Scale Set separated by operating system: `LinuxVirtualMachine`, `WindowsVirtualMachine`, `LinuxVirtualMachineScaleSet`, `WindowsVirtualMachineScaleSet`. The old-style generic `VirtualMachine` and `VirtualMachineScaleSet` are still available, so you can take your time before migrating existing stacks.

Here is a snippet that defines an Ubuntu VM:

``` ts
const vm = new azure.compute.LinuxVirtualMachine("server-vm", {
    resourceGroupName,
    networkInterfaceIds: [networkInterface.id],
    size: "Standard_A0",
    sourceImageReference: {
        publisher: "Canonical",
        offer: "UbuntuServer",
        sku: "16.04-LTS",
        version: "latest",
    },
    osDisk: {
        caching: "ReadWrite",
        storageAccountType: "Standard_LRS",
    },
    computerName: "hostname",
    adminUsername: username,
    adminPassword: password,
    disablePasswordAuthentication: false,
);
```

## Removing Deprecated Resources, Invokes, and Fields

> Azure Active Directory has its [own Pulumi provider](https://github.com/pulumi/pulumi-azuread/), so all the AD resources were now removed from the `AD` namespace of the Azure provider 2.0.

Reviewer: ```suggestion
Azure Active Directory now has its [own Pulumi provider](https://github.com/pulumi/pulumi-azuread/), so all the AD resources were removed from the `AD` namespace of the Azure Provider 2.0.
```

---
title: "Announcing Pulumi Azure Provider 2.0"
authors: ["mikhail-shilkov"]
tags: ["Azure"]

> date: "2020-03-11"

Reviewer 2: ```suggestion
date: "2020-03-12"
```

meta_desc: "Announcing the 2.0 release of Pulumi Azure provider and what's new in it."
---

> We are happy to announce the release of a new major version of Pulumi Azure provider. Pulumi Azure 2.0 is based on the [2.0 release](https://www.terraform.io/docs/providers/azurerm/guides/2.0-upgrade-guide.html) of the upstream provider and brings several improvements and breaking changes.

Reviewer: "a new major version of **the** Pulumi Azure provider."

> We are happy to announce the release of a new major version of Pulumi Azure provider. Pulumi Azure 2.0 is based on the [2.0 release](https://www.terraform.io/docs/providers/azurerm/guides/2.0-upgrade-guide.html) of the upstream provider and brings several improvements and breaking changes.

Reviewer: Missing the article 'the':
"a new major version of **the** Pulumi Azure provider."

Here is a snippet that defines an Ubuntu VM:

``` ts
const vm = new azure.compute.LinuxVirtualMachine("server-vm", {
    resourceGroupName,
    networkInterfaceIds: [networkInterface.id],
    size: "Standard_A0",
    sourceImageReference: {
        publisher: "Canonical",
        offer: "UbuntuServer",
        sku: "16.04-LTS",
        version: "latest",
    },
    osDisk: {
        caching: "ReadWrite",
        storageAccountType: "Standard_LRS",
    },
    computerName: "hostname",
    adminUsername: username,
    adminPassword: password,
    disablePasswordAuthentication: false,
);
```

## Removing Deprecated Resources, Invokes, and Fields

Azure Active Directory has its [own Pulumi provider](https://github.com/pulumi/pulumi-azuread/), so all the AD resources were now removed from the `AD` namespace of the Azure provider 2.0.

> A bunch of other resources, invokes, and fields were removed too, following the changes in the upstream provider. You can see the full list in [this upgrade guide](https://www.terraform.io/docs/providers/azurerm/guides/2.0-upgrade-guide.html).

Reviewer 2: ```suggestion
A number of other resources, invokes, and fields were removed, following the changes in the upstream provider. You can see the full list in [this upgrade guide](https://www.terraform.io/docs/providers/azurerm/guides/2.0-upgrade-guide.html).
```

## Default Versions in Callback Functions

> [Serverless as Callbacks](https://www.pulumi.com/blog/serverless-as-simple-callbacks-with-pulumi-and-azure-functions/) running on Azure Functions have bumped the default version of Azure Functions runtime to `~3` and `Node.js` to `~12`. As always, you can override the default to set the versions that you require.

Reviewer: Missing Article **the** below:
```suggestion
[Serverless as Callbacks](https://www.pulumi.com/blog/serverless-as-simple-callbacks-with-pulumi-and-azure-functions/) running on Azure Functions have bumped the default version of the Azure Functions runtime to `~3` and `Node.js` to `~12`. As always, you can override the default to set the versions that you require.
```

Callback Functions also moved from `ZipBlob` to `Blob` deploy as described above, so expect another replacement operation while upgrading.

## Custom Timeouts

Some types of resources take longer to create than others. For instance, provisioning a new Cosmos DB account may take from 10 minutes to over an hour depending on the number of geo locations and other factors.

> If you find that creation of a target resource times out, you can override the default timeout using resource options:

Reviewer 2: ```suggestion
If the creation of a target resource times out, you can override the default timeout using resource options:
```

``` ts
const cosmosdbAccount = new azure.cosmosdb.Account("cosmosdb", {
    resourceGroupName: resourceGroup.name,
    offerType: "Standard",
    geoLocations: manyLocations,
    consistencyPolicy: { consistencyLevel: "Session" },
}, {
    customTimeouts: {
        create: "30m",
        update: "60m",
    },
});
```

You shouldn't need this ability often, but it can be a lifesaver when you do.

## Getting Started with Migration

> To get started with Azure provider 2.0, update the versions in your package manager:

Reviewer 3: Would it make sense to use the language chooser here?

Author: Great idea!

```
// Node.js: package.json
"@pulumi/azure": "^2.0.0",

// .NET: csproj/fsproj/vbproj
<PackageReference Include="Pulumi.Azure" Version="2.0.0-preview" />

// Python: requirements.txt
pulumi-azure>=2.0.0

// Go: Gopkg.toml
[constraint]]
  version = "v2.0.0"
  name = "github.com/pulumi/pulumi-azure"
```

> There are quite a few breaking changes, so you may need to adjust your code before you would be able to run the program successfully again.

Reviewer 3: Anything notable we ought to call out? Or maybe link again to the TF upgrade guide?

Author: Added another link to the upgrade guide. Does it look better now?

> There are quite a few breaking changes, so you may need to adjust your code before you would be able to run the program successfully again.

Reviewer 2: ```suggestion
There are quite a few breaking changes, so you may need to adjust your code before you can run the program successfully again.
```

To help with this process, we upgraded all our Azure [templates](https://github.com/pulumi/templates/) and [examples](https://github.com/pulumi/examples/) to 2.0.

> If you want to stick with 1.x version for a while, please pin your version to `^1.0.0` or `1.14.0` in the package manager.

Reviewer 2: ```suggestion
If you want to stay with the 1.x version, please pin your version to `^1.0.0` or `1.14.0` in the package manager.
```

> If you still have an issue or a question, reach out to us on [Pulumi Community Slack](https://slack.pulumi.com/), and we'll help you through migration.

Reviewer 2: ```suggestion
If you have an issues or questions, reach out to us on [Pulumi Community Slack](https://slack.pulumi.com/), and we'll help you through migration.
```
